### PR TITLE
JsonHandler

### PR DESCRIPTION
**Pull request by levyishai**

### CODE REVIEW

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import edu.wpi.first.wpilibj.Filesystem;
+
+import java.io.*;
+
+public class JsonHandler {
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+    private static final String path = Filesystem.getDeployDirectory().getPath() + "\\";
+
+    /**
+     * Saves an object to a new or existing Json file.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Not "an object", saves _the_ object, the given object.
JSON.
You're forgetting to mention that there's parsing happening as well.

Comment (laviRZ) [reply src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Also mention that this uses the backup and temp method.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * @param object the object to save
+     * @param name   the name of the file to save to or create
+     */
+    public static void saveToJsonFile(Object object, String name) {
+        saveFile(JsonHandler.path + name + ".tmp", object);
+        renameFile(JsonHandler.path + name, JsonHandler.path + name + ".bak");
+        renameFile(JsonHandler.path + name + ".tmp", JsonHandler.path + name);
+        deleteFile(JsonHandler.path + name + ".bak");
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: extract this procedure to another method, `safeWrite` or something.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    /**
+     * Gets an object from a Json file.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Use the word "parsing".
JSON.
You don't "get a file", you get its name and then you read it.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * @param fileName the name of the file to get from
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: to read, to parse to whatever, not to "get from".

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     * @param type     the type of the object to get (a class)
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: parse.
You don't have to specify the type of the parameter type, it's already known.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+     * @return the object
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: parse

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import edu.wpi.first.wpilibj.Filesystem;
+
+import java.io.*;
+
+public class JsonHandler {
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+    private static final String path = Filesystem.getDeployDirectory().getPath() + "\\";
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: should be all caps

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
+package frc.trigon.robot.utilities;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import edu.wpi.first.wpilibj.Filesystem;
+
+import java.io.*;
+
+public class JsonHandler {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java:8]: move all file related functions to a new class `FilesHandler`

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Parses the given object to a new or existing JSON file,
+     * using a safe way of saving. (creating a temporary file and then
+     * replacing the original file with the temporary one)
+     *
+     * @param object the object to save
+     * @param name   the name of the file to save to or create
+     */
+    public static void saveToJsonFile(Object object, String name) throws IOException {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Don't use the word "save". "write" maybe.
And in the function name you need to explain that you're not getting text, but an object that you later parse to json
So maybe `parseToJsonAndWrite`

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import edu.wpi.first.wpilibj.Filesystem;
+
+import java.io.*;
+
+public class JsonHandler {
+    private static final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+    private static final String path = Filesystem.getDeployDirectory().getPath() + "\\";
+
+    /**
+     * Parses the given object to a new or existing JSON file,
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: You don't parse it to a file, you parse it to JSON. You then write the JSON to a new or existing file.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+     * using a safe way of saving. (creating a temporary file and then
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Safe way of writing. And you don't have to explain it here.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+    /**
+     * Parsing a JSON file to an object.
+     *
+     * @param fileName the name of the file to read from
+     * @param type     the type of the object to parse
+     * @return the parse
+     */
+    public static <T> T parseJsonToObject(String fileName, Class<T> type) {
+        try(Reader reader = new FileReader(JsonHandler.path + fileName)) {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Drop the parentheses, and move to a separate function.

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * @param object the object to save
+     * @param name   the name of the file to save to or create
+     */
+    public static void saveToJsonFile(Object object, String name) throws IOException {
+        safeWrite(parseObjectToJson(object), name);
+    }
+
+    /**
+     * Parsing a JSON file to an object.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Parses

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+     *
+     * @param fileName the name of the file to read from
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: To read, not read from

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+     * @param type     the type of the object to parse
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: the class?

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+     * @return the parse
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: the parse?

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+        writeStringToFile(JsonHandler.path + name + ".tmp", text);
+        renameFile(JsonHandler.path + name, JsonHandler.path + name + ".bak");
+        renameFile(JsonHandler.path + name + ".tmp", JsonHandler.path + name);
+        deleteFile(JsonHandler.path + name + ".bak");
+    }
+
+    private static void renameFile(String oldPath, String newPath) {
+        File file = new File(oldPath);
+        if(!file.renameTo(new File(newPath))) {
+            System.out.println("Failed to rename file.");
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Throw an exception

File: src/main/java/frc/trigon/robot/utilities/JsonHandler.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+        }
+    }
+
+    private static void deleteFile(String path) {
+        File file = new File(path);
+        if(!file.delete()) {
+            System.out.println("Failed to delete file.");
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/JsonHandler.java]: Throw an exception

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+
+import edu.wpi.first.wpilibj.Filesystem;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+
+public class FilesHandler {
+
+    private static final String PATH = Filesystem.getDeployDirectory().getPath() + "\\";
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: deploy path

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    /**
+     * Deletes a file.
+     *
+     * @param name the file name
+     * @throws IOException if the method failed to delete the specified file
+     */
+    public static void deleteFile(String name) throws IOException {
+        File file = new File(PATH + name);
+        if(!file.delete()) {
+            throw new IOException("Failed to delete file.");
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: add the file name

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+        FileWriter fileWriter = new FileWriter(PATH + name);
+        fileWriter.write(text);
+        fileWriter.close();
+    }
+
+    /**
+     * Renames a file.
+     *
+     * @param oldName the name of the file to rename
+     * @param newName the new name of the desired name
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: `name` and `newName`

**Review by laviRZ — CHANGES_REQUESTED** [review]
make `PATH` public, and make all the functions receive the "absolute path", with is the path plus the file name.

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+     * Creates a file using a safe method of writing.
+     * This method will write the string to a temporary file,
+     * rename the existing file to its name.bak.
+     * rename the temporary file to the desired name.
+     * delete the .bak file.
+     *
+     * @param name the name of the file to create
+     * @throws IOException if the method failed to safe write the file
+     */
+    public static void safeWrite(String text, String name) throws IOException {
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: switch around always have name before text

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+        FileReader reader;
+        StringBuilder fileContent = new StringBuilder();
+        reader = new FileReader(!doesFileExist(name) && doesFileExist(name + ".tmp") ? name + ".tmp" : name);
+        while(reader.read() != -1) {
+            fileContent.append((char) reader.read());
+        }
+        return fileContent.toString();
+    }
+
+    private static boolean doesFileExist(String name) {
```
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: Add docs

**Review by lycaon7 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java (hunk 23 above)
Comment (lycaon7) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: change name to `fileExist`

**Review by lycaon7 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java (hunk 23 above)
Comment (levyishai) [reply src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: Are you sure? LaviRZ told me to name it doesFileExist

Comment (levyishai) [reply src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: I did it anyway

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+
+public class FilesHandler {
+
+    public static final String DEPLOY_PATH = Filesystem.getDeployDirectory().getPath() + "\\";
+
+    /**
+     * Deletes a file.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: Not just _a_ file, the given file

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+     *
+     * @param name the file name
+     * @throws IOException if the method failed to delete the specified file
+     */
+    public static void deleteFile(String name) throws IOException {
+        File file = new File(name);
+        if(!file.delete()) {
+            throw new IOException("Failed to delete the file \"" + name + "\".");
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: Why the `'\'`?

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    /**
+     * Writes the given text to a file.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: "String" is more suitable than "text"

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+     * @throws IOException if the method failed to read the file
+     */
+    public static String readFile(String absolutePath) throws IOException {
+        String newPath = !fileExist(absolutePath) && fileExist(absolutePath + ".tmp") ? absolutePath + ".tmp" : absolutePath;
+        return Files.readString(Path.of(newPath));
+    }
+
+    private static boolean fileExist(String absolutePath) {
+        return new File(absolutePath).exists();
+    }
```
Comment (shay124578) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java:90]: Missing newline

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+    private static String extractPathFromAbsolutePath(String absolutePath) {
+        List<String> absolutePathSplit = new ArrayList<>(List.of(absolutePath.split("\\\\")));
+        absolutePathSplit.remove(absolutePathSplit.size() - 1);
+        return String.join("\\", absolutePathSplit) + "\\";
+    }
```
Comment (shay124578) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java:94]: Missing newline

**Review by shay124578 — DISMISSED** [review]

**Review by lycaon7 — DISMISSED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 29]
```diff
@@ -0,0 +1,103 @@
...
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+public class FilesHandler {
+
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: extra newline

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+        File file = new File(absolutePath);
+        if(!file.renameTo(new File(extractPathFromAbsolutePath(absolutePath) + newName))) {
+            throw new IOException("Failed to rename file " + absolutePath + " to " + newName);
+        }
+    }
+
+    /**
+     * Creates a file using a safe method of writing.
+     * This method will write the string to a temporary file,
+     * rename the existing file to its name.bak.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: comma at the end

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+     * rename the temporary file to the desired name.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: comma at the end

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+     * delete the .bak file.
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: and delete

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     * @param absolutePath the name of the file to create
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: to write to

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 34]
```diff
@@ -0,0 +1,103 @@
...
+     * @param text the text to write in the file
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: str not text

Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: write to the file not in

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 35]
```diff
@@ -0,0 +1,103 @@
...
+     * @param absolutePath the name of the file to read
+     * @return the file content
+     * @throws IOException if the method failed to read the file
+     */
+    public static String readFile(String absolutePath) throws IOException {
+        String newPath = !fileExist(absolutePath) && fileExist(absolutePath + ".tmp") ? absolutePath + ".tmp" : absolutePath;
+        return Files.readString(Path.of(newPath));
+    }
+
+    private static boolean fileExist(String absolutePath) {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: exists

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 36]
```diff
@@ -0,0 +1,105 @@
...
+    /**
+     * Renames a file.
+     *
+     * @param absolutePath the absolute path of the file to rename
+     * @param newName      the new name of the desired name
+     * @throws IOException if the method failed to rename the file
+     */
+    public static void renameFile(String absolutePath, String newName) throws IOException {
+        File file = new File(absolutePath);
+        if(!file.renameTo(new File(extractPathFromAbsolutePath(absolutePath) + newName))) {
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: extract this line to variables

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 37]
```diff
@@ -0,0 +1,99 @@
...
+    public static void deleteFile(String absolutePath) throws IOException {
+        File file = new File(absolutePath);
+        if(!file.delete())
+            throw new IOException("Failed to delete the file \"" + absolutePath + "\".");
+    }
+
+    /**
+     * Writes the given string to a file.
+     *
+     * @param absolutePath the files absolute path
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: file's

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 38]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Renames a file.
+     *
+     * @param absolutePath the absolute path of the file to rename
+     * @param newName      the new desired name
+     * @throws IOException if the method failed to rename the file
+     */
+    public static void renameFile(String absolutePath, String newName) throws IOException {
+        File file = new File(absolutePath);
+        String filePath = extractPathFromAbsolutePath(absolutePath);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: make the string be newAbsolutePath

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 39]
```diff
@@ -0,0 +1,99 @@
...
+        if(fileExists(absolutePath + ".bak"))
+            deleteFile(absolutePath + ".bak");
+    }
+
+    /**
+     * Reads a file and returns its content as a string.
+     * If the files does not exist, it will check for a .tmp file.
+     *
+     * @param absolutePath the absolute path of the file to read
+     * @return the file content
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: contents

**Review by laviRZ — CHANGES_REQUESTED** [review]

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 40]
```diff
@@ -0,0 +1,99 @@
...
+    public static void deleteFile(String absolutePath) throws IOException {
+        File file = new File(absolutePath);
+        if(!file.delete())
+            throw new IOException("Failed to delete the file \"" + absolutePath + "\".");
+    }
+
+    /**
+     * Writes the given string to a file.
+     *
+     * @param absolutePath the file's absolute path
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java:29]: every time you have `absolutePath` you should have the same explanation for it, in all functions

File: src/main/java/frc/trigon/robot/utilities/FilesHandler.java [hunk 41]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Renames a file.
+     *
+     * @param absolutePath the absolute path of the file to rename
+     * @param newName      the new desired name
+     * @throws IOException if the method failed to rename the file
+     */
+    public static void renameFile(String absolutePath, String newName) throws IOException {
+        File file = new File(absolutePath);
+        String newAbsolutePath = extractPathFromAbsolutePath(absolutePath);
```
Comment (laviRZ) [inline src/main/java/frc/trigon/robot/utilities/FilesHandler.java]: the name is wrong
this is the absolute path of the old file, not the new one

**Review by laviRZ — CHANGES_REQUESTED** [review]

**Review by lycaon7 — APPROVED** [review]

**Review by laviRZ — APPROVED** [review]

**Pull request merged**